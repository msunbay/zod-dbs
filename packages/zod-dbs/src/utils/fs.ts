import { existsSync, readdirSync, unlinkSync } from 'node:fs';
import { mkdir } from 'node:fs/promises';
import path from 'node:path';
import { logDebug } from 'zod-dbs-core';

import type { ZodDbsTableType } from 'zod-dbs-core';

/**
 * Deletes all .ts files in the output schemas folder.
 */
export function clearTablesDirectory(outputPath: string) {
  // Check if the directory exists first
  if (!existsSync(outputPath)) {
    logDebug(`Directory ${outputPath} does not exist, nothing to clear`);
    return;
  }

  const files = readdirSync(outputPath, {
    recursive: true,
    withFileTypes: true,
  });

  for (const file of files) {
    if (file.isDirectory()) continue;
    if (!file.name.endsWith('.ts')) continue;

    const filePath = path.join(file.path, file.name);
    unlinkSync(filePath);
  }

  if (files.length > 0) {
    logDebug(`Deleted all .ts files in ${outputPath}`);
  }
}

export async function ensureFolder(folderPath: string) {
  const exists = existsSync(folderPath);

  if (!exists) {
    await mkdir(folderPath, { recursive: true });
  }
}

export const getOutputFolder = (type: ZodDbsTableType): string => {
  switch (type) {
    case 'table':
      return 'tables';
    case 'materialized_view':
      return 'materialized_views';
    case 'view':
      return 'views';
    case 'foreign_table':
      return 'foreign_tables';
    default:
      return 'others';
  }
};
