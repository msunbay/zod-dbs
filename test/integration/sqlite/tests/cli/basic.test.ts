import fs from 'node:fs/promises';
import path from 'node:path';

import {
  executeCli,
  getOutputFiles,
  getProviderOutputDir,
} from '../../../utils/cli.js';
import { getProviderConfig } from '../../../utils/context.js';

it('CLI generates correct zod schemas with basic options', async () => {
  const config = getProviderConfig();
  const outputDir = getProviderOutputDir('sqlite', 'basic');

  await executeCli(
    `--provider sqlite --database "${config.database}" --schema-name "${config.schemaName}" --include "users" --output-dir "${outputDir}" --module-resolution esm`
  );

  const outputFiles = await getOutputFiles(outputDir);

  expect(outputFiles.length).toBe(5);

  for (const file of outputFiles) {
    const content = await fs.readFile(file, 'utf8');
    expect(content).toMatchSnapshot(path.relative(outputDir, file));
  }
});
