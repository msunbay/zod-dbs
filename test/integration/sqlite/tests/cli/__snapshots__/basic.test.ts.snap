// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CLI generates correct zod schemas with basic options > constants.ts 1`] = `
"// This file is auto-generated by zod-dbs. Do not edit this file directly.

export const TABLE_USERS = 'users';
"
`;

exports[`CLI generates correct zod schemas with basic options > tables/index.ts 1`] = `
"// This file is auto-generated by zod-dbs. Do not edit this file directly.

export * from './users/index.js';
"
`;

exports[`CLI generates correct zod schemas with basic options > tables/users/index.ts 1`] = `
"// This file is auto-generated by zod-dbs. Do not edit this file directly.

export * from './schema.js';
"
`;

exports[`CLI generates correct zod schemas with basic options > tables/users/schema.ts 1`] = `
"// This file is auto-generated by zod-dbs. Do not edit this file directly.

import { z } from 'zod';

export const USER_STATUSES = ['active', 'inactive', 'banned'] as const;

/**
 * Base read schema for the "users" table.
 * Validates raw rows read from the database (no casing transforms applied yet).
 */
export const UsersTableBaseSchema = z.object({
     /**
      * dataType: integer
      */
    id: z.number().nullable().transform((value) => value ?? undefined).optional(),
     /**
      * dataType: text
      */
    name: z.string(),
     /**
      * dataType: text
      */
    email: z.string().nullable().transform((value) => value ?? undefined).optional(),
     /**
      * dataType: text
      */
    status: z.enum(USER_STATUSES).nullable().transform((value) => value ?? undefined).optional(),
     /**
      * dataType: json
      */
    profile: z.any().nullable().transform((value) => value ?? undefined).optional(),
     /**
      * dataType: text
      */
    created_at: z.string().nullable().transform((value) => value ?? undefined).optional(),
});

/**
 * The base record type for the "users" table.
 * This type represents the raw database record without case transforms.
 */
export type UserBaseRecord = z.output<typeof UsersTableBaseSchema>;

/**
 * Read transform for the "users" table.
 * Maps database fields to app case properties.
 */
export const transformUserBaseRecord = (data: UserBaseRecord): {
    id?: UserBaseRecord['id'],
    name: UserBaseRecord['name'],
    email?: UserBaseRecord['email'],
    status?: UserBaseRecord['status'],
    profile?: UserBaseRecord['profile'],
    createdAt?: UserBaseRecord['created_at'],
} => ({
    id: data.id,
    name: data.name,
    email: data.email,
    status: data.status,
    profile: data.profile,
    createdAt: data.created_at,
});

/**
 * Read schema for the "users" table (after casing transform).
 */
export const UsersTableSchema = UsersTableBaseSchema.transform(transformUserBaseRecord);

/**
 * Base insert/write schema for the "users" table (no casing transforms).
 */
export const UsersTableInsertBaseSchema = z.object({
     /**
      * dataType: text
      */
    name: z.string(),
     /**
      * dataType: text
      */
    email: z.string().nullable().optional(),
     /**
      * dataType: text
      * defaultValue: 'active'
      */
    status: z.enum(USER_STATUSES).nullable().optional(),
     /**
      * dataType: json
      */
    profile: z.any().nullable().transform((value) => value ? JSON.stringify(value) : value).optional(),
     /**
      * dataType: text
      * defaultValue: datetime('now')
      */
    createdAt: z.string().nullable().optional(),
});

/**
 * The base record type for the "users" table.
 * This type represents an insertable database record before casing transformations are applied.
 */
export type UserInsertBaseRecord = z.output<typeof UsersTableInsertBaseSchema>;

/**
 * The base record type for the "users" table.
 * This type represents an updatable database record before casing transformations are applied.
 */
export type UserUpdateBaseRecord = Partial<UserInsertBaseRecord>;

/**
 * Insert transform for the "users" table.
 * Maps app cased properties to database fields.
 */
export const transformUserInsertBaseRecord = (data: UserInsertBaseRecord): {
    name: UserInsertBaseRecord['name'],
    email?: UserInsertBaseRecord['email'],
    status?: UserInsertBaseRecord['status'],
    profile?: UserInsertBaseRecord['profile'],
    created_at?: UserInsertBaseRecord['createdAt'],
} => ({
    name: data.name,
    email: data.email,
    status: data.status,
    profile: data.profile,
    created_at: data.createdAt,
});

/**
 * Update transform for the "users" table.
 * Maps app cased properties to database fields.
 */
export const transformUserUpdateBaseRecord = (data: UserUpdateBaseRecord): {
    name: UserUpdateBaseRecord['name'],
    email?: UserUpdateBaseRecord['email'],
    status?: UserUpdateBaseRecord['status'],
    profile?: UserUpdateBaseRecord['profile'],
    created_at?: UserUpdateBaseRecord['createdAt'],
} => ({
    name: data.name,
    email: data.email,
    status: data.status,
    profile: data.profile,
    created_at: data.createdAt,
});

/**
 * Insert schema for the "users" table (after casing transform).
 */
export const UsersTableInsertSchema = UsersTableInsertBaseSchema.transform(transformUserInsertBaseRecord);

/**
 * Update schema for the "users" table (after casing transform).
 */
export const UsersTableUpdateSchema = UsersTableInsertBaseSchema.partial().transform(transformUserUpdateBaseRecord);

type TableInsertRecord = z.input<typeof UsersTableInsertSchema>;
type TableReadRecord = z.output<typeof UsersTableSchema>;

/**
 * Read record (casing transformed) for the "users" table.
 */
export interface UserRecord {
    id?: TableReadRecord['id'];
    name: TableReadRecord['name'];
    email?: TableReadRecord['email'];
    status?: TableReadRecord['status'];
    profile?: TableReadRecord['profile'];
    createdAt?: TableReadRecord['createdAt'];
}

/**
 * Insert record (casing transformed) for the "users" table.
 */
export interface UserInsertRecord {
    name: TableInsertRecord['name'];
    email?: TableInsertRecord['email'];
    /**
    * @default: 'active'
    */
    status?: TableInsertRecord['status'];
    profile?: TableInsertRecord['profile'];
    /**
    * @default: datetime('now')
    */
    createdAt?: TableInsertRecord['createdAt'];
}

/**
 * Updatable record (casing transformed) for the "users" table.
 */
export type UserUpdateRecord = Partial<UserInsertRecord>;

export type UserStatus = (typeof USER_STATUSES)[number];
"
`;

exports[`CLI generates correct zod schemas with basic options > types.ts 1`] = `
"// This file is auto-generated by zod-dbs. Do not edit this file directly.

export type Table = 
    | 'users'
"
`;
